sdks/java/io/synthetic/src/main/java/org/apache/beam/sdk/io/synthetic/SyntheticStep.java

```diff
@@ -0,0 +1,203 @@
+
+      // Ensure the worker throughput is within the limit (if set).
+      if (options.maxWorkerThroughput < 0
+          || rateLimiterCache.getUnchecked(idAndThroughput).tryAcquire()) {
+        break;
+      } else { // try 1 millis extra delay
+        delayMsec = Duration.millis(1);
+      }
+    }
+
+    if (options.reportThrottlingMicros && sleepMillis > 0) {
+      throttlingMsecsCounter.inc(TimeUnit.MILLISECONDS.toMicros(sleepMillis));
+    }
+
+    if (options.preservesInputKeyDistribution) {
+      // Generate the new byte array value whose hashcode will be
+      // used as seed to initialize a Random object in next stages.
+      byte[] newValue = new byte[inputVal.length];
+      rnd.nextBytes(newValue);
+      return KV.of(inputKey, newValue);
+    } else {
+      return options.genKvPair(seed);
+    }
+  }
+
+  @ProcessElement
+  public void processElement(ProcessContext c) throws Exception {
+    byte[] key = c.element().getKey();
+    byte[] val = c.element().getValue();
+    int decimalPart = (int) options.outputRecordsPerInputRecord;
+    double fractionalPart = options.outputRecordsPerInputRecord - decimalPart;
+    // Use the hashcode of val as seed to make the test deterministic.
+    long hashCodeOfVal = options.hashFunction.hashBytes(val).asLong();
+    Random random = new Random(hashCodeOfVal);
+    int i;
+    for (i = 0; i < decimalPart; i++) {
+      c.output(outputElement(key, val, hashCodeOfVal, i, random));
+    }
+    if (random.nextDouble() < fractionalPart) {
+      c.output(outputElement(key, val, hashCodeOfVal, i, random));
+    }
+  }
+
+  @StartBundle
+  public void startBundle() throws Exception {
+    if (options.perBundleDelay > 0) {
+      SyntheticUtils.delay(
+          Duration.millis(options.perBundleDelay),
+          options.cpuUtilizationInMixedDelay,
+          options.perBundleDelayType,
+          new Random());
+    }
+  }
+
+  /**
+   * Synthetic step options. These options are all JSON, see documentations of individual fields for
+   * details. {@code Options} uses jackson annotations which PipelineOptionsFactory can use to parse
+   * and construct an instance.
+   */
+  public static class Options extends SyntheticOptions {
+    /**
+     * Amplification/filtering ratio: the number of output records should be emitted on average for
+     * each input record.
+     */
+    @JsonProperty public double outputRecordsPerInputRecord;
+
+    /**
+     * If false, the DoFn generates a different distribution of KV pairs according to the parameters
+     * in {@link SyntheticOptions}, and input records are merely used as a “clock”; If true, the
+     * shape of the input distribution is preserved, and the DoFn only does sleeping and
+     * amplification/filtering.
+     */
+    @JsonProperty public boolean preservesInputKeyDistribution;
+
+    /** User-defined tag to annotate synthetic step name. */
+    @JsonProperty public String uniqueTag;
```

Comment:
It is unused. Is it deliberate?

Reply:
I think this was used in an internal google version and is no longer useful.